### PR TITLE
ASoC: Intel: sof_ssp_amp: do not create amp link for nocodec board

### PR DESCRIPTION
**Pull request by brentlu**
A BE DAI link for speaker amplifier is always created even a board quirk specifies there is no amplifier. Modify the driver to check amplifier type before creating corresponding DAI link.

### CODE REVIEW

**Comment by brentlu** [comment]
Hi @CBala21, board rpl_lt6911_hdmi_ssp may suffer from this problem. Would you check the topology part? Thanks.

**Comment by CBala21** [comment]
> Hi @CBala21, board rpl_lt6911_hdmi_ssp may suffer from this problem. Would you check the topology part? Thanks.

Hi @brentlu , yes. we have reserved the dai link index for SSP codec in topology file even its not used in no codec mode.
now due to this changes it will break adl and rpl nocodec platforms which doesn't have SSP codec but still uses HDMI in capture and display audio. let us check with your code changes and update you. Thanks.

**Comment by brentlu** [comment]
Oh, if it's intentional, we could remove the link but keep BE id increased so we don't need to touch topologies. How does it sound?

**Comment by CBala21** [comment]
> Oh, if it's intentional, we could remove the link but keep BE id increased so we don't need to touch topologies. How does it sound?
@brentlu : yes. it will avoid changes on topology. as my understanding previous case also no DAI link created from driver side in nocodec case, if below flags are not set.
	if (sof_ssp_amp_quirk & SOF_RT1308_SPEAKER_AMP_PRESENT) {
		sof_rt1308_dai_link(&links[id]);
	} else if (sof_ssp_amp_quirk & SOF_CS35L41_SPEAKER_AMP_PRESENT) {
		cs35l41_set_dai_link(&links[id]);
	}
and its just adding one link id for SSP codec. because we have disabled DAI link creation from topology side (more over control from topology file) and no more devices will be enumerated for SSP codec amplifier. please correct me if I'm wrong.

Currently single topology source file used to build both SSP codec +HDMI-In and nocodec SSP+HDMI-in based on cmake flags. Hence dai link id is reserved and fixed.

**Comment by bardliao** [comment]
> Oh, if it's intentional, we could remove the link but keep BE id increased so we don't need to touch topologies. How does it sound?

Yes, keep BE id is necessary. I am wondering can we use predefined BE id for each type of dai link?
Something like
[#define SDW_JACK_OUT_DAI_ID 0](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw_common.h#L19)
[*be_id = codec_info->dais[dai_index].dailink[stream];](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L1442)

File: sound/soc/intel/boards/sof_ssp_amp.c [hunk 1]
```diff
@@ -231,41 +231,47 @@ sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
...
-	links[id].no_pcm = 1;
-	links[id].cpus = &cpus[id];
-	links[id].num_cpus = 1;
-	links[id].cpus->dai_name = devm_kasprintf(dev, GFP_KERNEL, "SSP%d Pin", ssp_codec);
-	if (!links[id].cpus->dai_name)
-		return NULL;
-
-	id++;
+	/* hdmi-in topology preserves this id with or without amplifier attached */
+	be_id++;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_ssp_amp.c]: What if `if (sof_ssp_amp_quirk & SOF_SSP_HDMI_CAPTURE_PRESENT)` is true? Won't be_id be added twice in the `if (sof_ssp_amp_quirk & SOF_SSP_HDMI_CAPTURE_PRESENT)` condition and here?

**Comment by brentlu** [comment]
I checked the sof-tgl-rt1308-hdmi-ssp.m4 and realized that the BE ID for dai link is fixed number. I modified machine driver part accordingly so the BE ID should match.

**Review by plbossart — APPROVED** [review]
not fully understanding the issue but the code looks ok to me.

@bardliao @CBala21 please review

**Review by bardliao — APPROVED** [review]

**Review by CBala21 — APPROVED** [review]

File: sound/soc/intel/boards/sof_ssp_amp.c [hunk 2]
```diff
@@ -222,7 +232,7 @@ sof_card_dai_links_create(struct device *dev, enum sof_ssp_codec amp_type,
 			links[id].name = devm_kasprintf(dev, GFP_KERNEL, "SSP%d-HDMI", port);
 			if (!links[id].name)
 				return NULL;
-			links[id].id = id;
+			links[id].id = fixed_be ? (HDMI_IN_BE_ID + i - 1) : id;
```
Comment (marc-hb) [inline sound/soc/intel/boards/sof_ssp_amp.c:235]: @brentlu can you explain when `fixed_be` is false here? It's defined `true` 15 lines above and never changed again.

Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_amp.c:235]: fixed_be will be false when the board config is 'adl_cs35l41'. Other board configs which implements HDMI-In are using the topology sof-tgl-rt1308-hdmi-ssp.m4 which is using fixed number for each BE link. I think there is still room for the code here to improve...

Comment (marc-hb) [reply sound/soc/intel/boards/sof_ssp_amp.c:235]: Thanks for submitting follow-up https://github.com/thesofproject/linux/pull/4694 so quickly. Let's move all discussions there.

**Pull request merged**